lib/rules/selector-anb-no-unmatchable/index.js

```diff
@@ -0,0 +1,117 @@
+'use strict';
+
+const { parse } = require('css-tree');
+const {
+	aNPlusBNotationPseudoClasses,
+	aNPlusBOfSNotationPseudoClasses,
+} = require('../../reference/selectors');
+const isStandardSyntaxRule = require('../../utils/isStandardSyntaxRule');
+
+const report = require('../../utils/report');
+const ruleMessages = require('../../utils/ruleMessages');
+const validateOptions = require('../../utils/validateOptions');
+
+const ruleName = 'selector-anb-no-unmatchable';
+
+const messages = ruleMessages(ruleName, {
+	rejected: (pseudoClass) => `Unexpected unmatchable An+B selector "${pseudoClass}"`,
+	rejectedParseError: (selector) => `Cannot parse selector "${selector}"`,
+});
+
+const meta = {
+	url: 'https://stylelint.io/user-guide/rules/selector-anb-no-unmatchable',
+};
+
+function isUnmatchableNth(/** @type {import('css-tree').AnPlusB} */ nth) {
+	const { a, b } = nth;
+
+	if (a !== null && a !== '0' && a !== '-0') {
+		return false;
+	}
+
+	if (b !== null && b !== '0' && b !== '-0') {
+		return false;
+	}
+
+	return true;
+}
+
+/** @type {import('stylelint').Rule} */
+const rule = (primary) => {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, { actual: primary });
+
+		if (!validOptions) {
+			return;
+		}
+
+		root.walkRules((ruleNode) => {
+			if (!isStandardSyntaxRule(ruleNode)) {
+				return;
+			}
+
+			ruleNode.selectors.forEach((selector) => {
+				/** @type {import('css-tree').CssNode} */
+				let cssTreeSelector;
+
+				try {
+					cssTreeSelector = parse(selector, { context: 'selector', positions: true });
+					checkSelector(cssTreeSelector);
+				} catch (e) {
+					report({
+						message: messages.rejectedParseError(ruleNode.selector),
+						node: ruleNode,
+						index: 0,
+						endIndex: 0,
+						result,
+						ruleName,
+					});
+				}
+			});
+
+			function checkSelector(/** @type {import('css-tree').CssNode} */ selector) {
+				if (selector.type !== 'Selector') {
+					return;
+				}
+
+				selector.children.forEach((selectorChild) => {
+					if (
+						selectorChild.type !== 'PseudoClassSelector' ||
+						(!aNPlusBNotationPseudoClasses.has(selectorChild.name) &&
+							!aNPlusBOfSNotationPseudoClasses.has(selectorChild.name))
+					) {
+						return;
+					}
+
+					const pseudoClassSelector = selectorChild;
+
+					if (pseudoClassSelector.children === null) {
+						return;
+					}
+
+					pseudoClassSelector.children.forEach((child) => {
+						if (child.type !== 'Nth' || child.nth.type !== 'AnPlusB') {
+							return;
+						}
+
+						if (isUnmatchableNth(child.nth)) {
+							report({
+								message: messages.rejected(pseudoClassSelector.name),
+								node: ruleNode,
+								index: pseudoClassSelector.loc?.start.column,
```

Comment:
question

Is `undefined` allowed as an index?

Reply:
I don't think so; there's a "soft" guarantee that if we include `positions` in the parse call, then `loc` is guaranteed to exist. Is there a better way for me to encode this? Ex, should I throw a parse error when `loc` doesn't exist?

Reply:
Even if passing `undefined` to `index` and `endIndex`, I think there is no inconvenience. Because the lint problem location (`column` and `endColumn` will point to the whole selector. Also, it seems a rare case.